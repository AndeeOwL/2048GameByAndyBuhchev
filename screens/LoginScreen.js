import AsyncStorage from "@react-native-async-storage/async-storage";
import { useNavigation } from "@react-navigation/native";
import { useState } from "react";
import { Alert, StyleSheet, Text, View } from "react-native";
import { Colors } from "../components/common/Colors";
import { Fonts } from "../components/common/Fonts";
import LoginButtons from "../components/Login/LoginButtons";
import LoginForm from "../components/Login/LoginForm";
import MainLogo from "../components/MainLogo";
import GameScreen from "./GameScreen";
import LeaderboardScreen from "./LeaderboardScreen";
import SignUpScreen from "./SignUpScreen";

function LoginScreen() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const navigation = useNavigation();

  const login = async () => {
    try {
      const user = AsyncStorage.getItem(username);
      const userObJ = AsyncStorage.getItem(JSON.parse(user));
      if (user !== null && userObJ.password === password) {
        setUsername("");
        setPassword("");
        navigation.navigate(GameScreen);
      } else {
        Alert.alert("Invalid input", "Invalid username or password");
      }
    } catch (error) {
      Alert.alert("Invalid input", "Invalid username or password");
    }
  };

  const signUp = () => {
    navigation.navigate(SignUpScreen);
  };

  const leaderboard = () => {
    navigation.navigate(LeaderboardScreen);
  };

  const usernameInputHandler = (enteredText) => {
    setUsername(enteredText);
  };

  const passwordInputHandler = (enteredText) => {
    setPassword(enteredText);
  };

  return (
    <View style={styles.rootContainer}>
      <MainLogo cStyle={styles.logo} />
      <Text style={styles.text}>Login to play</Text>
      <LoginForm
        usernameChange={usernameInputHandler}
        passwordChange={passwordInputHandler}
      />
      <LoginButtons login={login} signUp={signUp} leaderboard={leaderboard} />
    </View>
  );
}

export default LoginScreen;

const styles = StyleSheet.create({
  rootContainer: {
    flex: 1,
    backgroundColor: Colors.color256,
    alignItems: "center",
  },
  text: {
    color: Colors.darkBrown,
    fontSize: Fonts.big,
    fontWeight: "bold",
  },
});
