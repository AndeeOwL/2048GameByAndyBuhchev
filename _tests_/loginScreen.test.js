import { render } from "@testing-library/react-native";
import { fireEvent } from "@testing-library/react-native/build";
import LoginScreen from "../screens/LoginScreen";

const mockedNavigate = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe("Tests LoginScreen component", () => {
  test("renders correctly", () => {
    const tree = render(<LoginScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("it should render right amount of elements", () => {
    const { getAllByText, getAllByPlaceholderText } = render(<LoginScreen />);
    const logo = getAllByText("2048");
    expect(logo).toHaveLength(1);
    const text = getAllByText("Login to play");
    expect(text).toHaveLength(1);
    const username = getAllByPlaceholderText("username");
    expect(username).toHaveLength(1);
    const password = getAllByPlaceholderText("password");
    expect(password).toHaveLength(1);
    const loginButton = getAllByText("Login");
    expect(loginButton).toHaveLength(1);
    const signUpButton = getAllByText("SignUp");
    expect(signUpButton).toHaveLength(1);
    const leaderboardButton = getAllByText("Leaderboard");
    expect(leaderboardButton).toHaveLength(1);
  });

  test("it should change username and password", () => {
    const { getByPlaceholderText } = render(<LoginScreen />);
    const username = getByPlaceholderText("username");
    const password = getByPlaceholderText("password");

    fireEvent.changeText(username, "andrean");
    fireEvent.changeText(password, "12345678");
  });

  test("Pressed login with invalid inputs should not navigate", () => {
    const { getByText } = render(<LoginScreen />);
    const login = getByText("Login");
    fireEvent.press(login);
    expect(mockedNavigate).toHaveBeenCalledTimes(0);
  });

  test("Pressed signUp should navigate", () => {
    const { getByText } = render(<LoginScreen />);
    const signUp = getByText("SignUp");
    fireEvent.press(signUp);
    expect(mockedNavigate).toHaveBeenCalledTimes(1);
  });

  test("Pressed login with invalid inputs should not navigate", () => {
    const { getByText } = render(<LoginScreen />);
    const leaderboard = getByText("Leaderboard");
    fireEvent.press(leaderboard);
    expect(mockedNavigate).toHaveBeenCalledTimes(2);
  });
});
