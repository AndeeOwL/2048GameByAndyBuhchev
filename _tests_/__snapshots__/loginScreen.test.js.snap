// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests LoginScreen component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fae873",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5db31",
        "height": 150,
        "justifyContent": "center",
        "marginBottom": 50,
        "marginTop": 50,
        "width": 150,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 35,
          "fontWeight": "bold",
        }
      }
    >
      2048
    </Text>
  </View>
  <Text
    style={
      Object {
        "color": "#4d4646",
        "fontSize": 32,
        "fontWeight": "bold",
      }
    }
  >
    Login to play
  </Text>
  <View>
    <TextInput
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="username"
      style={
        Array [
          Object {
            "backgroundColor": "#ded6d5",
            "fontSize": 24,
            "height": 50,
            "marginBottom": 15,
            "marginLeft": 15,
            "marginRight": 15,
            "marginTop": 15,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
            "width": 300,
          },
        ]
      }
    />
    <TextInput
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="password"
      style={
        Array [
          Object {
            "backgroundColor": "#ded6d5",
            "fontSize": 24,
            "height": 50,
            "marginBottom": 15,
            "marginLeft": 15,
            "marginRight": 15,
            "marginTop": 15,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
            "width": 300,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        onPress={[Function]}
        style={
          Object {
            "backgroundColor": "#f5a164",
            "fontSize": 16,
            "fontWeight": "bold",
            "margin": 15,
            "padding": 15,
          }
        }
      >
        Login
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        onPress={[Function]}
        style={
          Object {
            "backgroundColor": "#f5a164",
            "fontSize": 16,
            "fontWeight": "bold",
            "margin": 15,
            "padding": 15,
          }
        }
      >
        SignUp
      </Text>
    </View>
  </View>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        onPress={[Function]}
        style={
          Object {
            "backgroundColor": "#f5a164",
            "fontSize": 16,
            "fontWeight": "bold",
            "margin": 15,
            "padding": 15,
          }
        }
      >
        Leaderboard
      </Text>
    </View>
  </View>
</View>
`;
